src/Evolution/Systems/Ccz4/TimeDerivative.cpp

```diff
@@ -0,0 +1,604 @@
+        d_conformal_christoffel_second_kind,  // eq 16
+    const gsl::not_null<tnsr::Ijj<DataVector, Dim>*>
+        christoffel_second_kind,  // eq 17
+    const gsl::not_null<tnsr::ij<DataVector, Dim>*>
+        spatial_ricci_tensor_buffer,  // buffer for eq 18 -20
+    const gsl::not_null<tnsr::ii<DataVector, Dim>*>
+        spatial_ricci_tensor,  // eq 18 - 20
+    const gsl::not_null<tnsr::ij<DataVector, Dim>*> grad_grad_lapse,  // eq 21
+    const gsl::not_null<Scalar<DataVector>*> divergence_lapse,        // eq 22
+    const gsl::not_null<tnsr::I<DataVector, Dim>*>
+        contracted_conformal_christoffel_second_kind,  // eq 23
+    const gsl::not_null<tnsr::iJ<DataVector, Dim>*>
+        d_contracted_conformal_christoffel_second_kind,  // eq 24
+    const gsl::not_null<tnsr::i<DataVector, Dim>*>
+        spatial_z4_constraint,  // eq 25
+    const gsl::not_null<Scalar<DataVector>*>
+        upper_spatial_z4_constraint_buffer,  // buffer for eq 25
+    const gsl::not_null<tnsr::I<DataVector, Dim>*>
+        upper_spatial_z4_constraint,  // eq 25
+    const gsl::not_null<tnsr::ij<DataVector, Dim>*>
+        grad_spatial_z4_constraint,  // eq 26
+    const gsl::not_null<Scalar<DataVector>*>
+        ricci_scalar_plus_divergence_z4_constraint,  // eq 27
+    // free params
+    const double c, const double cleaning_speed,  // e
+    const Scalar<DataVector>& eta, const double f,
+    const Scalar<DataVector>& k_0, const tnsr::i<DataVector, Dim>& d_k_0,
+    const double kappa_1, const double kappa_2, const double kappa_3,
+    const double mu, const double one_over_relaxation_time,  // \tau^{-1}
+    const bool evolve_shift,                                 // s
+    const bool use_harmonic_slicing_condition,  // g(\alpha)
+    // evolved variables
+    const tnsr::ii<DataVector, Dim>& conformal_spatial_metric,
+    const Scalar<DataVector>& ln_lapse, const tnsr::I<DataVector, Dim>& shift,
+    const Scalar<DataVector>& ln_conformal_factor,
+    const tnsr::ii<DataVector, Dim>& a_tilde,
+    const Scalar<DataVector>& trace_extrinsic_curvature,
+    const Scalar<DataVector>& theta, const tnsr::I<DataVector, Dim>& gamma_hat,
+    const tnsr::I<DataVector, Dim>& b, const tnsr::i<DataVector, Dim>& field_a,
+    const tnsr::iJ<DataVector, Dim>& field_b,
+    const tnsr::ijj<DataVector, Dim>& field_d,
+    const tnsr::i<DataVector, Dim>& field_p,
+    // spatial derivatives of evolved variables
+    const tnsr::ijj<DataVector, Dim>& d_a_tilde,
+    const tnsr::i<DataVector, Dim>& d_trace_extrinsic_curvature,
+    const tnsr::i<DataVector, Dim>& d_theta,
+    const tnsr::iJ<DataVector, Dim>& d_gamma_hat,
+    const tnsr::iJ<DataVector, Dim>& d_b,
+    const tnsr::ij<DataVector, Dim>& d_field_a,
+    const tnsr::ijK<DataVector, Dim>& d_field_b,
+    const tnsr::ijkk<DataVector, Dim>& d_field_d,
+    const tnsr::ij<DataVector, Dim>& d_field_p) {
+  constexpr double one_third = 1.0 / 3.0;
+  constexpr double eulers_number = 2.71828182845904523536;
```

Comment:
`<cmath>` offers a `std::exp()` function, however, I don't know if it is preferable over the constexpr/pow combo.

Reply:
Thanks! I defined the literal just so I could have it be `constexpr`, but I can imagine that `exp` might be specialized and faster. So, I'm also not sure what's preferred!

Reply:
glibc, at least, calculates `pow(x, y)` as `exp(y * log(x))`, so `exp` is probably faster and more accurate.  (Also don't hardcode math constants.  There is `M_E` in `<cmath>`.)

Reply:
Thank you both! I'll make this change